{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pre-process DecryptM Dataset\n",
    "\n",
    "**Publication**: Jana Zecha et al. Decrypting drug actions and protein modifications by dose- and time-resolved proteomics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import toml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dose-Dependent Drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_files(directory):\n",
    "    \"\"\"\n",
    "    Searches for .txt and .toml files in the given directory and its subdirectories.\n",
    "    Args:\n",
    "        directory (str): The path to the directory to search in.\n",
    "    Returns:\n",
    "        tuple: A tuple containing two lists:\n",
    "            - txt_files (list): A list of .txt files found.\n",
    "            - toml_files (list): A list of .toml files found.\n",
    "    \"\"\"\n",
    "    txt_files = []\n",
    "    toml_files = []\n",
    "\n",
    "    for root, dirs, files in os.walk(directory):\n",
    "        for file in files:\n",
    "            if file.endswith('.txt'):\n",
    "                txt_files.append(file)\n",
    "            elif file.endswith('.toml'):\n",
    "                toml_files.append(file)\n",
    "\n",
    "    return txt_files, toml_files\n",
    "\n",
    "\n",
    "def process_experiment_column(df):\n",
    "    \"\"\"\n",
    "    Processes the 'Experiment' column in the given DataFrame by splitting it into multiple new columns.\n",
    "    Args:\n",
    "        df (pd.DataFrame): The input DataFrame containing an 'Experiment' column to be processed.\n",
    "    Returns:\n",
    "        pd.DataFrame: The DataFrame with the 'Experiment' column split into new columns:\n",
    "            - 'Experiment'\n",
    "            - 'Cell Line'\n",
    "            - 'Drug'\n",
    "            - 'Time point'\n",
    "            - 'Replicate'\n",
    "    \"\"\"\n",
    "    new_columns = [\"Experiment\", \"Cell Line\", \"Drug\", \"Time point\", \"Replicate\"]\n",
    "    split_columns = df['Experiment'].str.split('_', expand=True)\n",
    "    num_parts = split_columns.shape[1]\n",
    "    \n",
    "    for col in new_columns:\n",
    "        df[col] = 'NA'\n",
    "    \n",
    "    df[new_columns[0]] = split_columns[0]\n",
    "    df[new_columns[1]] = split_columns[1]\n",
    "    df[new_columns[2]] = split_columns.iloc[:, 2:num_parts-2].apply(lambda x: '_'.join(x), axis=1)\n",
    "    df[new_columns[3]] = split_columns[num_parts-2]\n",
    "    df[new_columns[4]] = split_columns[num_parts-1]\n",
    "    \n",
    "    df['Time point'] = df['Time point'].apply(lambda x: int(x[:-1]) * 60 if x.endswith('h') else int(x[:-3]) if x.endswith('min') else x)  # Convert hours to minutes and make column numeric\n",
    "    \n",
    "    df = df.loc[:, ~df.columns.duplicated()]\n",
    "    df = df[new_columns + [col for col in df.columns if col not in new_columns]]\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "def update_dataframe_with_toml(df, toml_file):\n",
    "    \"\"\"\n",
    "    Updates the column names of a DataFrame based on information from a TOML file.\n",
    "    This function reads a TOML file to get dose and channel information, then renames\n",
    "    the columns of the DataFrame accordingly. The columns to be renamed are expected\n",
    "    to follow the pattern \"TMT Channel Ratio {channel}\", and they will be renamed to\n",
    "    \"Dose {dose}\" based on the corresponding dose for each channel.\n",
    "    Args:\n",
    "        df (pandas.DataFrame): The DataFrame whose columns are to be renamed.\n",
    "        toml_file (str): The path to the TOML file containing dose and channel information.\n",
    "    Returns:\n",
    "        pandas.DataFrame: The DataFrame with updated column names.\n",
    "    \"\"\"\n",
    "    with open(toml_file, 'r') as file:\n",
    "        toml_data = toml.load(file)\n",
    "    doses = toml_data['TMT']['doses']\n",
    "    channels = toml_data['TMT']['channels']\n",
    "    \n",
    "    for i, channel in enumerate(channels):\n",
    "        old_column_name = f\"TMT Channel Ratio {channel}\"\n",
    "        if old_column_name in df.columns:\n",
    "            new_column_name = f\"Dose {doses[i]}\" if i < len(doses) else f\"Dose NA\"\n",
    "            df.rename(columns={old_column_name: new_column_name}, inplace=True)\n",
    "    return df\n",
    "\n",
    "\n",
    "def filter_data(df, min_score_cutoff = 60, max_pep_cutoff = 0.05, verbose = False):\n",
    "    \"\"\"\n",
    "    Filters the DataFrame based on the given cutoffs.\n",
    "    \n",
    "    Args:\n",
    "        df (pd.DataFrame): The input DataFrame to be filtered.\n",
    "        max_score_cutoff (float): The maximum score cutoff for filtering.\n",
    "        min_pep_cutoff (float): The minimum PEP cutoff for filtering.\n",
    "    \n",
    "    Returns:\n",
    "        pd.DataFrame: The filtered DataFrame.\n",
    "    \"\"\"\n",
    "    rows_before = df.shape[0]\n",
    "\n",
    "    filtered_df = df[\n",
    "        (df['Max Score'] >= min_score_cutoff) & \n",
    "        (df['Min PEP'] <= max_pep_cutoff) & \n",
    "        (df['Phospho (STY)'] >= 1) & \n",
    "        (df['Phosphoproteome'] == True)\n",
    "    ]\n",
    "\n",
    "    rows_after = filtered_df.shape[0]\n",
    "    if verbose == True:\n",
    "        print(f\"Number of rows before filtering: {rows_before}\")\n",
    "        print(f\"Number of rows after filtering: {rows_after}\")\n",
    "    \n",
    "    return filtered_df\n",
    "\n",
    "\n",
    "def remove_columns(df):\n",
    "    columns_to_remove = [\n",
    "        'Experiment', 'Replicate', 'N duplicates', 'Sequence', 'Length',\n",
    "        'Missed cleavages', 'Proteins', 'Leading proteins', 'Protein names',\n",
    "        'Phospho (STY)', 'All Phospho (STY) Probabilities', 'Max Score', 'Min PEP',\n",
    "        'Intensity', 'Phosphoproteome', 'Fullproteome', 'Curve signal', 'Log EC50',\n",
    "        'Curve slope', 'Curve top', 'Curve bottom', 'R2', 'Curve RMSE', 'Log EC50 error',\n",
    "        'Curve slope error', 'Curve top error', 'Curve bottom error', 'EC50', 'pEC50',\n",
    "        'Curve effect size', 'Regulation', 'Acetyl (K)', 'All Acetyl (K) Probabilities', 'Acetylproteome'\n",
    "    ]\n",
    "\n",
    "    df.drop(columns=columns_to_remove, inplace=True, errors='ignore')\n",
    "    \n",
    "    # Remove columns that start with 'Reporter intensity corrected' or 'TMT Channel Normal'\n",
    "    df = df.loc[:, ~df.columns.str.startswith(('Reporter intensity corrected', 'TMT Channel Normal'))]\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "def pivot_long(df):\n",
    "    \"\"\"\n",
    "    Converts a DataFrame from wide format to long format.\n",
    "    This function takes a DataFrame with multiple columns representing different doses\n",
    "    and pivots it to a long format where each row represents a single dose measurement.\n",
    "    Args:\n",
    "        df (pd.DataFrame): The input DataFrame in wide format.\n",
    "    Returns:\n",
    "        pd.DataFrame: The transformed DataFrame in long format with columns 'Dose' and 'Value'.\n",
    "    \"\"\"\n",
    "    \n",
    "    dose_columns = [col for col in df.columns if col.startswith('Dose')]\n",
    "    \n",
    "    df_long = pd.melt(\n",
    "        df, \n",
    "        id_vars=[col for col in df.columns if col not in dose_columns], \n",
    "        value_vars=dose_columns, \n",
    "        var_name='Dose', \n",
    "        value_name='Value'\n",
    "    )\n",
    "    \n",
    "    df_long['Dose'] = df_long['Dose'].str.replace('Dose ', '').astype(float)\n",
    "    \n",
    "    return df_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cell Line</th>\n",
       "      <th>Drug</th>\n",
       "      <th>Time point</th>\n",
       "      <th>Modified sequence</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Dose</th>\n",
       "      <th>Value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A549</td>\n",
       "      <td>PD325901</td>\n",
       "      <td>60</td>\n",
       "      <td>(ac)AAAAAAAGDS(ph)DSWDADAFSVEDPVR</td>\n",
       "      <td>EIF3J</td>\n",
       "      <td>Dose 10000.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A549</td>\n",
       "      <td>PD325901</td>\n",
       "      <td>60</td>\n",
       "      <td>(ac)AAAAAAAGDS(ph)DSWDADAFSVEDPVRK</td>\n",
       "      <td>EIF3J</td>\n",
       "      <td>Dose 10000.0</td>\n",
       "      <td>1.489595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A549</td>\n",
       "      <td>PD325901</td>\n",
       "      <td>60</td>\n",
       "      <td>(ac)AAAADS(ph)FSGGPAGVR</td>\n",
       "      <td>RMI2</td>\n",
       "      <td>Dose 10000.0</td>\n",
       "      <td>1.307801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A549</td>\n",
       "      <td>MK2206</td>\n",
       "      <td>60</td>\n",
       "      <td>(ac)AAAAPDSRVS(ph)EEENLK</td>\n",
       "      <td>RRP15</td>\n",
       "      <td>Dose 10000.0</td>\n",
       "      <td>0.789177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A549</td>\n",
       "      <td>PD325901</td>\n",
       "      <td>60</td>\n",
       "      <td>(ac)AAAAPDSRVS(ph)EEENLK</td>\n",
       "      <td>RRP15</td>\n",
       "      <td>Dose 10000.0</td>\n",
       "      <td>1.065007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998965</th>\n",
       "      <td>HeLa</td>\n",
       "      <td>SAHA</td>\n",
       "      <td>240</td>\n",
       "      <td>YYS(ph)DSDDELTVEQR</td>\n",
       "      <td>BOD1L1</td>\n",
       "      <td>Dose 0.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998966</th>\n",
       "      <td>HeLa</td>\n",
       "      <td>SAHA</td>\n",
       "      <td>240</td>\n",
       "      <td>YYS(ph)IDDNQNK</td>\n",
       "      <td>NCOA7</td>\n",
       "      <td>Dose 0.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998967</th>\n",
       "      <td>HeLa</td>\n",
       "      <td>SAHA</td>\n",
       "      <td>240</td>\n",
       "      <td>YYS(ph)PCEEHPAETNQNEGSESGTIR</td>\n",
       "      <td>ARHGEF5</td>\n",
       "      <td>Dose 0.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998968</th>\n",
       "      <td>HeLa</td>\n",
       "      <td>SAHA</td>\n",
       "      <td>240</td>\n",
       "      <td>YYSDS(ph)DDELTVEQR</td>\n",
       "      <td>BOD1L1</td>\n",
       "      <td>Dose 0.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998969</th>\n",
       "      <td>HeLa</td>\n",
       "      <td>SAHA</td>\n",
       "      <td>240</td>\n",
       "      <td>YYVT(ph)IIDAPGHR</td>\n",
       "      <td>EEF1A1P5;EEF1A1</td>\n",
       "      <td>Dose 0.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9998970 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Cell Line      Drug  Time point                   Modified sequence  \\\n",
       "0            A549  PD325901          60   (ac)AAAAAAAGDS(ph)DSWDADAFSVEDPVR   \n",
       "1            A549  PD325901          60  (ac)AAAAAAAGDS(ph)DSWDADAFSVEDPVRK   \n",
       "2            A549  PD325901          60             (ac)AAAADS(ph)FSGGPAGVR   \n",
       "3            A549    MK2206          60            (ac)AAAAPDSRVS(ph)EEENLK   \n",
       "4            A549  PD325901          60            (ac)AAAAPDSRVS(ph)EEENLK   \n",
       "...           ...       ...         ...                                 ...   \n",
       "9998965      HeLa      SAHA         240                  YYS(ph)DSDDELTVEQR   \n",
       "9998966      HeLa      SAHA         240                      YYS(ph)IDDNQNK   \n",
       "9998967      HeLa      SAHA         240        YYS(ph)PCEEHPAETNQNEGSESGTIR   \n",
       "9998968      HeLa      SAHA         240                  YYSDS(ph)DDELTVEQR   \n",
       "9998969      HeLa      SAHA         240                    YYVT(ph)IIDAPGHR   \n",
       "\n",
       "              Gene names          Dose     Value  \n",
       "0                  EIF3J  Dose 10000.0       NaN  \n",
       "1                  EIF3J  Dose 10000.0  1.489595  \n",
       "2                   RMI2  Dose 10000.0  1.307801  \n",
       "3                  RRP15  Dose 10000.0  0.789177  \n",
       "4                  RRP15  Dose 10000.0  1.065007  \n",
       "...                  ...           ...       ...  \n",
       "9998965           BOD1L1      Dose 0.0  1.000000  \n",
       "9998966            NCOA7      Dose 0.0  1.000000  \n",
       "9998967          ARHGEF5      Dose 0.0  1.000000  \n",
       "9998968           BOD1L1      Dose 0.0  1.000000  \n",
       "9998969  EEF1A1P5;EEF1A1      Dose 0.0  1.000000  \n",
       "\n",
       "[9998970 rows x 7 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Cell Lines: ['A549' 'PC-9' 'RPMI8226' 'K562' 'A431' 'BT-474' 'HeLa' 'MDA-MB-175'\n",
      " 'A459' 'KYSE-520' 'SK-BR-3']\n",
      "Unique Drugs: ['PD325901' 'MK2206' 'AZD8055' 'Dactolisib' 'Dasatinib' 'Nintedanib'\n",
      " 'Tideglusib' 'Pictilisib' 'AZD4547' 'LapatinibAZD4547' 'Lapatinib'\n",
      " 'BTZ_CFZ' 'Imatinib' 'Pertuzumab' 'Trastuzumab' 'CUDC101' 'Curcumin'\n",
      " 'Cytarabine' 'Afatinib' 'Gefitinib' 'GeftinibAZD4547-1to80' 'A486' 'A485'\n",
      " 'Refametinib' 'Staursporin' 'Methotrexat' 'Paclitaxel' 'Romidepsin'\n",
      " 'SelumetinibMK2206-1to2' 'SelumetinibMK2206-3to1' 'Selumetinib' 'SHP099'\n",
      " 'SAHA']\n",
      "Unique Time Points: [ 60  30 960 120 240 480]\n",
      "Unique Doses: ['Dose 10000.0' 'Dose 3000.0' 'Dose 1000.0' 'Dose 300.0' 'Dose 100.0'\n",
      " 'Dose 30.0' 'Dose 10.0' 'Dose 3.0' 'Dose 1.0' 'Dose 0.0' 'Dose 0.3'\n",
      " 'Dose 0.1' 'Dose 0.03' 'Dose 0.01' 'Dose 0.003' 'Dose 100000.0'\n",
      " 'Dose 30000.0' 'Dose 0.001' 'Dose 0.0003' 'Dose 0.0001']\n"
     ]
    }
   ],
   "source": [
    "def main():\n",
    "    directory = '../data'\n",
    "    txt_files, toml_files = search_files(directory)\n",
    "    combined_df = pd.DataFrame()\n",
    "    \n",
    "    # Remove time-dependent file\n",
    "    if 'Rituximab_td.txt' in txt_files:\n",
    "        txt_files.remove('Rituximab_td.txt')\n",
    "        \n",
    "    for txt_file in txt_files:\n",
    "        txt_file_path = os.path.normpath(directory + '/' + txt_file)\n",
    "        df = pd.read_csv(txt_file_path, delimiter='\\t', header=0)\n",
    "        df = process_experiment_column(df)\n",
    "\n",
    "        toml_file_name = f\"pipeline_{txt_file.replace('_dd.txt', '.toml')}\"\n",
    "        if toml_file_name in toml_files:\n",
    "            toml_file_name = os.path.normpath(directory + '/' + toml_file_name)\n",
    "            df = update_dataframe_with_toml(df, toml_file_name)\n",
    "        else:\n",
    "            print(f\"toml file: {toml_file_name} does not exist!\")\n",
    "            \n",
    "        min_score_cutoff = 60\n",
    "        max_pep_cutoff = 0.05\n",
    "        df = filter_data(df, min_score_cutoff, max_pep_cutoff, False)\n",
    "        df = remove_columns(df)\n",
    "        df = pivot_long(df)\n",
    "        \n",
    "        combined_df = pd.concat([combined_df, df], ignore_index=True)\n",
    "    \n",
    "    return combined_df\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    combined_df = main()\n",
    "    display(combined_df)\n",
    "    \n",
    "    # Check unique values\n",
    "    unique_cell_lines = combined_df['Cell Line'].unique()\n",
    "    unique_drugs = combined_df['Drug'].unique()\n",
    "    unique_time_points = combined_df['Time point'].unique()\n",
    "    unique_doses = combined_df['Dose'].unique()\n",
    "    \n",
    "    print(\"Unique Cell Lines:\", unique_cell_lines)\n",
    "    print(\"Unique Drugs:\", unique_drugs)\n",
    "    print(\"Unique Time Points:\", unique_time_points)\n",
    "    print(\"Unique Doses:\", unique_doses)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rituximab Time-Dependent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_file_path = os.path.join(\"..\", \"data\", \"rituximab_td.txt\")\n",
    "\n",
    "rituximab_td = pd.read_csv(data_file_path, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter data\n",
    "# Arbitrary Cutoffs\n",
    "max_score_cutoff = 60  # Confidence score for peptide identification\n",
    "min_pep_cutoff = 0.05  # Minimum posterior error probability\n",
    "\n",
    "rows_before = rituximab_td.shape[0]\n",
    "print(f\"Number of rows before filtering: {rows_before}\")\n",
    "\n",
    "rituximab_td = rituximab_td[\n",
    "    (rituximab_td['Max Score'] >= max_score_cutoff) & \n",
    "    (rituximab_td['Min PEP'] <= min_pep_cutoff) &\n",
    "    (rituximab_td['Phospho (STY)'] >= 1) &\n",
    "    (rituximab_td['Phosphoproteome'] == True)\n",
    "]\n",
    "\n",
    "rows_after = rituximab_td.shape[0]\n",
    "print(f\"Number of rows before filtering: {rows_after}\")\n",
    "\n",
    "# Break down experiment column\n",
    "split_experiment_col = rituximab_td['Experiment'].str.split('_', expand=True)\n",
    "rituximab_td['Cell Line'] = split_experiment_col[1]\n",
    "rituximab_td['Drug'] = split_experiment_col[2]\n",
    "rituximab_td['Dose'] = split_experiment_col[3].str.replace('ng', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rituximab_td = rituximab_td[[\n",
    "        'Cell Line',\n",
    "        'Drug',\n",
    "        'Dose',\n",
    "        'Modified sequence', \n",
    "        'Gene names',\n",
    "        'TMT Channel Ratio 1',\n",
    "        'TMT Channel Ratio 2', \n",
    "        'TMT Channel Ratio 3', \n",
    "        'TMT Channel Ratio 4',\n",
    "        'TMT Channel Ratio 5', \n",
    "        'TMT Channel Ratio 6', \n",
    "        'TMT Channel Ratio 7',\n",
    "        'TMT Channel Ratio 8'\n",
    "]]\n",
    "\n",
    "rituximab_td = rituximab_td.rename(columns={\n",
    "        'TMT Channel Ratio 1': 'Ratio_1',\n",
    "        'TMT Channel Ratio 2': 'Ratio_2',\n",
    "        'TMT Channel Ratio 3': 'Ratio_5',\n",
    "        'TMT Channel Ratio 4': 'Ratio_10',\n",
    "        'TMT Channel Ratio 5': 'Ratio_60',\n",
    "        'TMT Channel Ratio 6': 'Ratio_120',\n",
    "        'TMT Channel Ratio 7': 'Ratio_360',\n",
    "        'TMT Channel Ratio 8': 'Ratio_1440'})\n",
    "\n",
    "rituximab_td.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "deeplearning",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
